smack-extensions/src/main/java/org/jivesoftware/smackx/avatar/UserAvatarManager.java

```diff
@@ -0,0 +1,318 @@
+import org.jivesoftware.smackx.pubsub.LeafNode;
+import org.jivesoftware.smackx.pubsub.PayloadItem;
+import org.jivesoftware.smackx.pubsub.PubSubException;
+
+/**
+ * User Avatar manager class.
+ *
+ * @author Fernando Ramirez
+ * @author Paul Schaub
+ * @see <a href="http://xmpp.org/extensions/xep-0084.html">XEP-0084: User
+ *      Avatar</a>
+ */
+public final class UserAvatarManager extends Manager {
+
+    public static final String DATA_NAMESPACE = "urn:xmpp:avatar:data";
+    public static final String METADATA_NAMESPACE = "urn:xmpp:avatar:metadata";
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    private static final Map<XMPPConnection, UserAvatarManager> INSTANCES = new WeakHashMap<>();
+
+    private final PepManager pepManager;
+
+    /**
+     * Get the singleton instance of UserAvatarManager.
+     *
+     * @param connection
+     * @return the instance of UserAvatarManager
+     */
+    public static synchronized UserAvatarManager getInstanceFor(XMPPConnection connection) {
+        UserAvatarManager userAvatarManager = INSTANCES.get(connection);
+
+        if (userAvatarManager == null) {
+            userAvatarManager = new UserAvatarManager(connection);
+            INSTANCES.put(connection, userAvatarManager);
+        }
+
+        return userAvatarManager;
+    }
+
+    private UserAvatarManager(XMPPConnection connection) {
+        super(connection);
+        this.pepManager = PepManager.getInstanceFor(connection);
+    }
+
+    /**
+     * Returns true if User Avatar is supported by the server.
+     *
+     * @return true if User Avatar is supported by the server.
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public boolean isSupportedByServer()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        return findNodeItem(DATA_NAMESPACE) && findNodeItem(METADATA_NAMESPACE);
+    }
+
+    private boolean findNodeItem(String nodeName)
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        ServiceDiscoveryManager serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection());
+        List<Item> items = serviceDiscoveryManager.discoverItems(connection().getUser().asBareJid(), nodeName)
+                .getItems();
+        return items != null && items.size() > 0;
+    }
+
+    /**
+     * Get the data node.
+     *
+     * @return the data node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getDataNode()
```

Comment:
I'd argue this method should be private.

Reply:
Make it so, we can later make it public if it is really needed. But a good API has as less as possible public methods.